File: control-plane/service-go/internal/auth/identity.go
package auth

import (
	"regexp"
	"strings"
)

var phonePattern = regexp.MustCompile(`^\+[1-9]\d{7,14}$`)

func NormalizeInternationalPhoneNumber(value string) string {
	normalized := strings.NewReplacer(" ", "", "\t", "", "\n", "", "(", "", ")", "", ".", "", "-", "").Replace(strings.TrimSpace(value))
	if !phonePattern.MatchString(normalized) {
		return ""
	}
	return normalized
}

func NormalizeAuthIdentifier(value string) string {
	if phone := NormalizeInternationalPhoneNumber(value); phone != "" {
		return phone
	}
	return strings.ToLower(strings.TrimSpace(value))
}

func GeneratedUsernameFromPhone(phoneNumber, userID string) string {
	digits := strings.TrimPrefix(phoneNumber, "+")
	prefix := userID
	if len(prefix) > 8 {
		prefix = prefix[:8]
	}
	return "phone_" + digits + "_" + prefix
}

File: control-plane/service/src/modules/auth/identity.ts
export function normalizeInternationalPhoneNumber(value: string) {
  const normalized = value.trim().replace(/[\s().-]/g, "");
  if (!/^\+[1-9]\d{7,14}$/.test(normalized)) {
    return null;
  }
  return normalized;
}

export function normalizeAuthIdentifier(value: string) {
  const phoneNumber = normalizeInternationalPhoneNumber(value);
  return phoneNumber ?? value.trim().toLowerCase();
}

export function generatedUsernameFromPhone(phoneNumber: string, userId: string) {
  const digits = phoneNumber.replace(/^\+/, "");
  return `phone_${digits}_${userId.slice(0, 8)}`;
}
